local-volume/utils/update-pv-to-beta/update_pv_na_to_beta.go

```diff
@@ -0,0 +1,132 @@
+)
+
+func cloneAndUpdatePV(pv *v1.PersistentVolume) (*v1.PersistentVolume, error) {
+	var affinity v1.NodeAffinity
+	err := json.Unmarshal([]byte(pv.Annotations[v1.AlphaStorageNodeAffinityAnnotation]), &affinity)
+	if err != nil {
+		return nil, err
+	}
+	pvClone := pv.DeepCopy()
+	delete(pvClone.Annotations, v1.AlphaStorageNodeAffinityAnnotation)
+	pvClone.Spec.NodeAffinity = &v1.VolumeNodeAffinity{
+		Required: affinity.RequiredDuringSchedulingIgnoredDuringExecution,
+	}
+	return pvClone, nil
+}
+func updateLocalPVAlphaAnn(client *kubernetes.Clientset, pv *v1.PersistentVolume) error {
+	// if it is not local PV, return directly
+	if pv.Spec.Local == nil {
+		return nil
+	}
+	// check alpha node affinity annotation
+	if len(pv.Annotations) > 0 && pv.Annotations[v1.AlphaStorageNodeAffinityAnnotation] != "" {
+		pvClone, err := cloneAndUpdatePV(pv)
+		if err != nil {
+			return err
+		}
+
+		glog.Infof("Updating local PV(%s), node affinity is: %v", pv.Name, pv.Annotations[v1.AlphaStorageNodeAffinityAnnotation])
+		_, err = client.CoreV1().PersistentVolumes().Update(pvClone)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				glog.Errorf("PV(%s) seems to have been deleted", pv.Name)
+				return nil
+			}
+			if errors.IsConflict(err) {
+				glog.Errorf("PV(%s) seems to have been updated, will get the latest resource version and retry", pv.Name)
+				return getAndUpdatePV(pv.Name, client)
+			}
+			return err
+		}
+	}
+	return nil
+}
+
+func getAndUpdatePV(pvName string, client *kubernetes.Clientset) error {
+	glog.Infof("Get latest local PV object")
+	pv, err := client.CoreV1().PersistentVolumes().Get(pvName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	pvClone, err := cloneAndUpdatePV(pv)
+	if err != nil {
+		return err
+	}
+	glog.Infof("Update PV(%S) again", pvName)
+	_, err = client.CoreV1().PersistentVolumes().Update(pvClone)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func startSwitchingLocalPVAlphaAnn(client *kubernetes.Clientset) error {
+	pvs, err := client.CoreV1().PersistentVolumes().List(metav1.ListOptions{})
+	if err != nil {
+		glog.Errorf("list PVs error: %v", err)
+	}
+	if len(pvs.Items) == 0 {
+		glog.Infof("No PVs found, return directly")
+		return nil
+	}
+
+	for _, pv := range pvs.Items {
+		if err = updateLocalPVAlphaAnn(client, &pv); err != nil {
+			glog.Errorf("update PV: %s error, %v", pv.Name, err)
+			return err
+		}
+	}
+	return nil
+}
+
+func main() {
+	flag.Set("logtostderr", "true")
+	flag.Parse()
+
+	client := common.SetupClient()
+
+	glog.Infof("Starting to update local PV node affinity to beta... ")
+	var err error
+	for i := 0; i < 5; i++ {
+		err = startSwitchingLocalPVAlphaAnn(client)
+		// if the err is nil, return directly;
+		// if not, no matter the err is TooManyRequests or others, retry several times
+		if err == nil {
+			return
+		}
+		time.Sleep(1 * time.Minute)
+	}
+	if err != nil {
+		glog.Errorf("update local PVs alpha node affinity to beta error, %v", err)
```

Comment:
Let's exit with an error code in this case.

Reply:
done

Reply:
Yes let's exit with an error.